### ISSUE TITLE
Aura utilities cache taint can break edit mode and actionbars

### ISSUE THREAD

**Issue by Meorawr** [Stanzilla/WoWUIBugs#390]
Various aura utility functions utilize cache tables to speed up repeated lookups. The entries in these caches can be tainted if insecure code indirectly calls these functions, leading to taint spreading via these cache tables, through the party/raid frames, and into edit mode - whereby the use of keybound abilities on actionbars can then break.

# Details

The problematic functions are as follows:

- [AuraUtil.ShouldDisplayDebuff(unitCaster, spellId)](https://github.com/Gethe/wow-ui-source/blob/de21439f9be0e4b03b5a489784b6343acea53525/Interface/FrameXML/AuraUtil.lua#L200)
  - This internally calls `GetCachedVisibilityInfo` which can read and write tainted entries into the `cachedVisualizationInfo` table.
- [AuraUtil.ShouldDisplayBuff(unitCaster, spellId, canApplyAura)](https://github.com/Gethe/wow-ui-source/blob/de21439f9be0e4b03b5a489784b6343acea53525/Interface/FrameXML/AuraUtil.lua#L218)
  - This internally calls `CheckIsSelfBuff` which can read and write tainted entries into the `cachedSelfBuffChecks` table.
- [AuraUtil.IsPriorityDebuff(spellId)](https://github.com/Gethe/wow-ui-source/blob/de21439f9be0e4b03b5a489784b6343acea53525/Interface/FrameXML/AuraUtil.lua#L239-L244)
  - This internally calls `CheckIsPriorityAura` which can read and write tainted entries into the `cachedPriorityChecks` table.

The primary vector through which taint in these cache tables manages to spill out everywhere is through the following sequence of events originating from the compact raid frame manager.

1. When players join or leave a group the `GROUP_ROSTER_UPDATE` event will be fired. This event is handled by the [CompactRaidFrameContainer](https://github.com/Gethe/wow-ui-source/blob/de21439f9be0e4b03b5a489784b6343acea53525/Interface/AddOns/Blizzard_CompactRaidFrames/Blizzard_CompactRaidFrameContainer.lua#L50-L52) to trigger an update of the raid frames.
2. When laying out the frames the [AddPlayers](https://github.com/Gethe/wow-ui-source/blob/de21439f9be0e4b03b5a489784b6343acea53525/Interface/AddOns/Blizzard_CompactRaidFrames/Blizzard_CompactRaidFrameContainer.lua#L147-L166) method is called which will add a unit frame for each group member to the container and set its unit via [CompactUnitFrame_SetUnit](https://github.com/Gethe/wow-ui-source/blob/de21439f9be0e4b03b5a489784b6343acea53525/Interface/FrameXML/CompactUnitFrame.lua#L164).
3. When the unit of a compact raid frame is configured, [CompactUnitFrame_UpdateAll](https://github.com/Gethe/wow-ui-source/blob/de21439f9be0e4b03b5a489784b6343acea53525/Interface/FrameXML/CompactUnitFrame.lua#L317-L336) is called which will lead to a call into [CompactUnitFrame_UpdateAuras](https://github.com/Gethe/wow-ui-source/blob/de21439f9be0e4b03b5a489784b6343acea53525/Interface/FrameXML/CompactUnitFrame.lua#L1382).
4. This leads to calls on multiple paths to [AuraUtil.ProcessAura](https://github.com/Gethe/wow-ui-source/blob/de21439f9be0e4b03b5a489784b6343acea53525/Interface/FrameXML/AuraUtil.lua#L131) for potentially every aura that the unit being assigned has. **This function can - on certain paths - call any of the previously mentioned problematic functions and read a previously-tainted cached entry from the internal tables, which will then taint execution.**
5. Steps 3 and 4 will be repeated for potentially all auras on every member in the group.
6. On the path back to the initial stack frame of the event handler, some tainted value is being persisted within the compact raid frame container.
7. When entering edit mode, [EditModeAccountSettingsMixin:OnEnter](https://github.com/Gethe/wow-ui-source/blob/de21439f9be0e4b03b5a489784b6343acea53525/Interface/FrameXML/EditModeManager.lua#L1508-L1519) will be triggered and will lead to function calls that refresh the raid and party frame displays. This seemingly reads a tainted field from step 6 at some point, tainting execution and spreading that into edit mode. This may not _immediately_ break edit mode, however it can be forced to do so if the user exits and then re-enters edit mode afterwards.

# Test case

This specific setup will make use of a permanent duration aura that the player can apply on-demand. For this we'll use the [Swift Red Gryphon](https://www.wowhead.com/spell=32289) mount aura with spell ID `32289`. Additionally, two accounts are recommended as we'll need to join a group to trigger the spread of taint.

1. Log into the game and ensure the player is not currently mounted or in a group. If you are mounted or grouped, leave the group, dismount, and reload the UI.
9. Run the following command: `/run AuraUtil.ShouldDisplayBuff("player", 32289, true)`
10. Summon the mount: `/cast Swift Red Gryphon`
11. Invite the second account to join the group.
12. After the second account has joined the group (such that two party/raid unit frames have appeared), enter Edit Mode.
13. Exit Edit Mode.
14. Re-enter Edit Mode. You'll get an action blocked error at this point.
15. Exit Edit Mode.
16. Cycle the current action bar page up and down once.
17. Attempt to use an ability on the action bars via a keybinding; you'll get another blocked action error.

# Solutions

The problematic functions can all be trivially fixed with one-line tweaks to route the internal calls through a [securecallfunction](https://wowpedia.fandom.com/wiki/API_securecallfunction) barrier - this would prevent taint from entries within the tables from leaking back out to callers, and should be safe from a security perspective as these tables themselves are not externally accessible in any way.

```diff
diff --git a/Interface/FrameXML/AuraUtil.lua b/Interface/FrameXML/AuraUtil.lua
index dd57486b..dfa27196 100644
--- a/Interface/FrameXML/AuraUtil.lua
+++ b/Interface/FrameXML/AuraUtil.lua
@@ -165,6 +165,8 @@ function AuraUtil.ProcessAura(aura, displayOnlyDispellableDebuffs, ignoreBuffs,
 end
 
 do
+	local securecallfunction = securecallfunction;
+
 	local hasValidPlayer = false;
 	EventRegistry:RegisterFrameEvent("PLAYER_ENTERING_WORLD");
 	EventRegistry:RegisterFrameEvent("PLAYER_LEAVING_WORLD");
@@ -198,7 +200,7 @@ do
 	end
 
 	function AuraUtil.ShouldDisplayDebuff(unitCaster, spellId)
-		local hasCustom, alwaysShowMine, showForMySpec = GetCachedVisibilityInfo(spellId);
+		local hasCustom, alwaysShowMine, showForMySpec = securecallfunction(GetCachedVisibilityInfo, spellId);
 		if ( hasCustom ) then
 			return showForMySpec or (alwaysShowMine and (unitCaster == "player" or unitCaster == "pet" or unitCaster == "vehicle") );	--Would only be "mine" in the case of something like forbearance.
 		else
@@ -216,7 +218,7 @@ do
 	end
 
 	function AuraUtil.ShouldDisplayBuff(unitCaster, spellId, canApplyAura)
-		local hasCustom, alwaysShowMine, showForMySpec = GetCachedVisibilityInfo(spellId);
+		local hasCustom, alwaysShowMine, showForMySpec = securecallfunction(GetCachedVisibilityInfo, spellId);
 	
 		if ( hasCustom ) then
 			return showForMySpec or (alwaysShowMine and (unitCaster == "player" or unitCaster == "pet" or unitCaster == "vehicle"));
@@ -238,11 +240,11 @@ do
 	if ( classFilename == "PALADIN" ) then
 		AuraUtil.IsPriorityDebuff = function(spellId)
 			local isForbearance = (spellId == 25771);
-			return isForbearance or CheckIsPriorityAura(spellId);
+			return isForbearance or securecallfunction(CheckIsPriorityAura, spellId);
 		end
 	else
 		AuraUtil.IsPriorityDebuff = function(spellId)
-			return CheckIsPriorityAura(spellId);
+			return securecallfunction(CheckIsPriorityAura, spellId);
 		end
 	end
 
```

**Comment by Meorawr** [comment]
Fixed in 10.0.7.47910.

**Comment by funkydude** [comment]
Also fixed in [10.0.5.47936](https://github.com/BigWigsMods/WoWUI/commit/8ee82005bd2e099ae803bde128628fb3886aea3d)
